jmdaemon/jmdaemon/onionmc.py

```diff
@@ -583,19 +583,23 @@ def try_to_connect(self) -> None:
 
 class OnionDirectoryPeer(OnionPeer):
     delay = 4.0
+
     def try_to_connect(self) -> None:
         # Delay deliberately expands out to very
         # long times as yg-s tend to be very long
         # running bots:
         self.delay *= 1.5
         if self.delay > 10000:
             log.warn("Cannot connect to directory node peer: {} "
-                     "after 20 attempts, giving up.".format(self.peer_location()))
+                     "after 20 attempts, giving up.".format(
+                         self.peer_location()))
             return
-        try:
-            self.connect()
-        except OnionPeerConnectionError:
-            reactor.callLater(self.delay, self.try_to_connect)
+        #try:
+        log.info(f"Going to reattempt connection to {self.peer_location()} in "
+                 f"{self.delay} seconds.")
+        # randomize by a few seconds to minimize bursty-ness locally
+        jitter = random.randint(-1, 5)
+        reactor.callLater(self.delay + jitter, self.connect)

```

Comment:
Question: Should we reset `self.delay` at some point? I suppose this ties into:

> I spent a lot of time trying to find a 'deeper' solution, looking for a way to distinguish between 'this was an attempt to start a connection to an onion service which somehow immediately aborted' and 'this was a preexisting connection that was working fine and has now dropped'

Would this mean that every time the directory disconnects, say once every few days because of natural/innocous causes, we would keep expanding the delay and eventually give up entirely?

Reply:
> Would this mean that every time the directory disconnects, say once every few days because of natural/innocous causes, we would keep expanding the delay and eventually give up entirely?

The delay is currently capped at 10k seconds, so a bit less than 3 hours.

It may well be possible to track 'how long have we been connected', and if that is a reasonably long interval, then reset `self.delay` back to 4s (or whatever). I'll just leave it for now though (if it ever got up to 10k seconds, then something was quite badly wrong at some point in the past, during the running of this bot instance ... until we're really sure we know what this problem is, and how to handle it, I don't want to be too opinionated or sophisticated .. the more logic I add the more chance something is wrong).

Reply:
Doh, what am I saying, it's not *capped* at 10k seconds, it just gives up at that point. I'm not sure why I did that, I will change it.

Reply:
Fixed that in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1436/commits/79eaf4ab99c0d18d5a66f2dc773aaca45654cec9 - it should now be as discussed above.